cupy/cuda/cufft.pyx

```diff
@@ -7,6 +7,18 @@ from cupy.cuda cimport memory
 from cupy.cuda cimport stream as stream_module
 
 
+cdef object _current_plan = None
```

Comment:
Could you store it in `threading.local()`?

Reply:
Good point. Just did. In the beginning I followed the approach in handling streams and devices:
```diff
diff --git a/cupy/cuda/cufft.pyx b/cupy/cuda/cufft.pyx
index 8180011..9b1b314 100644
--- a/cupy/cuda/cufft.pyx
+++ b/cupy/cuda/cufft.pyx
@@ -1,5 +1,6 @@
 cimport cython  # NOQA
 import numpy
+import threading
 
 import cupy
 from cupy.cuda cimport driver
@@ -7,7 +8,29 @@ from cupy.cuda cimport memory
 from cupy.cuda cimport stream as stream_module
 
 
-cdef object _current_plan = None
+cdef object _thread_local = threading.local()
+
+
+cdef class _PlanThreadLocal:
+    cdef object current_plan
+
+    def __cinit__(self):
+        self.current_plan = None
+
+    @staticmethod
+    cdef _PlanThreadLocal get():
+        plan = getattr(_thread_local, 'cufft_plan', None)
+        if plan is None:
+            tlp = _PlanThreadLocal()
+            _thread_local.cufft_plan = tlp
+            return tlp
+        return <_PlanThreadLocal>plan
+
+    cdef set_current_plan(self, plan):
+        self.current_plan = plan
+
+    cdef get_current_plan(self):
+        return self.current_plan
 
 
 cpdef get_current_plan():
@@ -16,7 +39,8 @@ cpdef get_current_plan():
     Returns:
         None or cupy.cuda.cufft.Plan1d or cupy.cuda.cufft.PlanNd
     """
-    return _current_plan
+    tlp = _PlanThreadLocal.get()
+    return tlp.get_current_plan()
 
 
 cdef extern from 'cupy_cufft.h' nogil:
@@ -126,13 +150,13 @@ class Plan1d(object):
         check_result(result)
 
     def __enter__(self):
-        global _current_plan
-        _current_plan = self
+        tlp = _PlanThreadLocal.get()
+        tlp.set_current_plan(self)
         return self
 
     def __exit__(self, exc_type, exc_value, traceback):
-        global _current_plan
-        _current_plan = None
+        tlp = _PlanThreadLocal.get()
+        tlp.set_current_plan(None)
 
     def fft(self, a, out, direction):
         cdef Handle plan = self.plan
@@ -265,13 +289,13 @@ class PlanNd(object):
         check_result(result)
 
     def __enter__(self):
-        global _current_plan
-        _current_plan = self
+        tlp = _PlanThreadLocal.get()
+        tlp.set_current_plan(self)
         return self
 
     def __exit__(self, exc_type, exc_value, traceback):
-        global _current_plan
-        _current_plan = None
+        tlp = _PlanThreadLocal.get()
+        tlp.set_current_plan(None)
 
     def fft(self, a, out, direction):
         cdef Handle plan = self.plan
```
but I realized it's an overkill for cuFFT plans, and so switched to the current implementation.